authenticator/authenticator.go

```diff
@@ -41,77 +84,122 @@ func NewAuthenticator() (*Authenticator, error) {
 
 func (a *Authenticator) run() {
 	for {
-		log.Printf("authenticator running. %d requests received\n", a.counter)
+		log.Printf("authenticator running. %d requests received", a.counter)
 		time.Sleep(10 * time.Second)
 	}
 }
 
-func IAMFormatToSTS(iamArn string) (string, error) {
-	matches := regexp.MustCompile(`arn:aws:iam::(.*):role/(.*)`).FindStringSubmatch(iamArn)
-	if matches == nil {
-		return "", fmt.Errorf("provided IAM role ARN is not properly formatted, expected format: arn:aws:iam::accountid:role/rolename")
-	}
-	accountId := matches[1]
-	role := matches[2]
-	return fmt.Sprintf("arn:aws:sts::%s:assumed-role/%s", accountId, role), nil
-}
-
 func executeGetCallerIdentity(request string) (string, error) {
 	resp, err := http.Post(request, "", nil)
 	if err != nil {
 		return "", err
 	}
-	responseData, err := ioutil.ReadAll(resp.Body)
+
+	// We can ignore errors here because if it's not a valid response body, we'll still error
+	// below when we try to unmarshal the XML.
+	respBody, _ := ioutil.ReadAll(resp.Body)
+	if resp.StatusCode != 200 {
+		return "", fmt.Errorf("received unexpected get caller identity response %d: %s", resp.StatusCode, respBody)
+	}
+
 	type GetCallerIdentityResponse struct {
 		IamArn string `xml:"GetCallerIdentityResult>Arn"`
 	}
 	response := GetCallerIdentityResponse{}
-	err = xml.Unmarshal(responseData, &response)
+	err = xml.Unmarshal(respBody, &response)
 	if err != nil {
 		return "", err
 	}
-	iamArn := strings.Trim(response.IamArn, "{}")
-	return iamArn, nil
+	return response.IamArn, nil
 }
 
-func verifyService(claimedIamArn, signedGetCallerIdentity string) error {
-	log.Printf("verifying service for role: %s\n", claimedIamArn)
-	actualIamArn, err := executeGetCallerIdentity(signedGetCallerIdentity)
+// verifyIdentity takes a signed get caller identity string and executes
+// the request to the given AWS STS endpoint. It returns the caller's
+// full IAM ARN.
+func verifyIdentity(signedGetCallerIdentity string) (string, error) {
```

Comment:
I am now a little unsure about the name of this function. Maybe "getIdentity" is more appropriate? I realize that it does some verification on the request, but I don't want it to be confused with verifying that the service is who it says it is.

Reply:
Sure! Updated in https://github.com/approzium/approzium/pull/47/commits/35c5d680018af10f8f1a9d819617a915787a037f.